### PR TITLE
Make c10::string_view an alias of std::string_view

### PR DESCRIPTION
**Pull request by cyyever**
In order to move from c10::string_view to std::string_view,  c10::string_view was renamed to c10::string_view_ext. Several places using starts_with or ends_with of  the old c10::string_view were changed to  use  the new name.
get_fully_qualified_type_name also  uses  c10::string_view_ext to reflect an accurate function signature. 
c10/test/util/string_view_test.cpp had been largely modified to use c10::string_view_ext in the assertions.

cc @albanD

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by clee2000** [comment]
@pytorchbot revert -m "breaking some internal tests, probably usages of string_view that need to be changed?" -c ghfirst

Here are some random bits of logs
```
E1014 23:25:34.291942 12014 ExceptionTracer.cpp:221] exception stack complete
terminate called after throwing an instance of 'c10::Error'
  what():  Type c10::basic_string_view<char> could not be converted to any of the known types.
Exception raised from operator() at ...ATen/core/jit_type.h:1810 (most recent call first):

# 2  c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
# 3  c10::detail::getTypePtr_<c10::basic_string_view<char> >::call()::{lambda()#1}::operator()() const
# 4  c10::Type::SingletonOrSharedTypePtr<c10::Type> c10::getFakeTypePtrCopy<c10::basic_string_view<char> >()
# 5  0x000000000058d7f7
# 6  c10::detail::infer_schema::make_function_schema(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&, c10::ArrayRef<c10::detail::infer_schema::ArgumentDef>, c10::ArrayRef<c10::detail::infer_schema::ArgumentDef>)
# 7  0x000000000003bba4
# 8  call_init.part.0
# 9  _dl_init
# 10 __GI__dl_catch_exception
# 11 dl_open_worker
# 12 __GI__dl_catch_exception
```

```
Stderr:
error: use of undeclared identifier 'str'; did you mean 'at::str'?
  158 |           tupleNames_.push_back(str(*it));
      |                                 ^~~
      |                                 at::str
c10/util/StringUtil.h:115:23: note: 'at::str' declared here
  115 | inline decltype(auto) str(const Args&... args) {
      |                       ^
1 error generated.
```

I'm going to try changing the str to at::str but idk if that's the correct thing to do

**Comment by cyyever** [comment]
@clee2000 Fixed, help import it to verify?

**Review by albanD — COMMENTED** [review]
Please do add details on why you added the skip-pr-sanity-check label and in general I wouldn't not expect you to add it without discussing it.
Int this particular case, this would ONLY be acceptable to skip it if you share the command you used to codemod the change you did. From a quick look, this covers a lot more than just making it an alias as the title of the PR suggests.

Please do add these details to the description if you do want to skip the pr sanity check tests.

**Comment by cyyever** [comment]
@albanD Thank you for your advice.  I will refine the description.

**Comment by albanD** [comment]
Thanks!
Would you be able to share a set of sed/codemod command we can run locally to re-generate the exact same diff without having to check every line?

**Comment by cyyever** [comment]
@albanD I managed to simplify the test file changes and rebased.

**Review by albanD — COMMENTED** [review]
Ho much easier to review, thanks!
I'll let @clee2000 check if this needs an internal equivalent.

**Comment by clee2000** [comment]
I'm seeing failures on the import like 
```
...\NNTorchScriptBackend.cpp(647): error C2440: '<function-style-cast>': cannot convert from 'initializer list' to 'fmt::v9::string_view'
    ...\NNTorchScriptBackend.cpp(647): note: 'fmt::v9::basic_string_view<char>::basic_string_view': no overloaded function could convert all the argument types
    ...\fmt\core.h(442): note: could be 'fmt::v9::basic_string_view<char>::basic_string_view(const Char *,size_t) noexcept'
            with
            [
                Char=char
            ]
    ...\NNTorchScriptBackend.cpp(647): note: 'fmt::v9::basic_string_view<char>::basic_string_view(const Char *,size_t) noexcept': cannot convert argument 1 from 'std::_String_view_iterator<_Traits>' to 'const Char *'
            with
            [
                Char=char
            ]
            and
            [
                _Traits=std::char_traits<char>
            ]
            and
            [
                Char=char
            ]
    ...\NNTorchScriptBackend.cpp(647): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
    ...\fmt\core.h(468): note: or       'fmt::v9::basic_string_view<char>::basic_string_view(S) noexcept'
    ...\fmt\core.h(462): note: or       'fmt::v9::basic_string_view<char>::basic_string_view(const std::basic_string<Char,Traits,Alloc> &) noexcept'
            with
            [
                Char=char
            ]
    ...\NNTorchScriptBackend.cpp(647): note: while trying to match the argument list '(std::_String_view_iterator<_Traits>, std::basic_string_view<char,std::char_traits<char>>::size_type)'
            with
            [
                _Traits=std::char_traits<char>
            ]
    ...\NNTorchScriptBackend.cpp(647): error C3615: constexpr function 'fmt::v9::basic_format_parse_context<char,fmt::v9::detail::error_handler>::on_error' cannot result in a constant expression
    ...\fmt\core.h(718): note: failure was caused by call of undefined function or one not declared 'constexpr'
    ...\fmt\core.h(718): note: see usage of 'fmt::v9::detail::error_handler::on_error'
    ...\NNTorchScriptBackend.cpp(647): note: the call stack of the evaluation (the oldest call first) is
    ...\NNTorchScriptBackend.cpp(647): note: while evaluating function 'fmt::v9::basic_format_string<char,const size_t &,std::basic_string<char,std::char_traits<char>,std::allocator<char>>>::basic_format_string<char[48],0>(const S (&))'
            with
            [
                S=char [48]
            ]
    ...\fmt\core.h(3169): note: while evaluating function 'void fmt::v9::detail::parse_format_string<true,char,checker>(fmt::v9::basic_string_view<char>,Handler &&)'
            with
            [
                Handler=checker
            ]
    ...\fmt\core.h(2731): note: while evaluating function 'const Char *fmt::v9::detail::parse_replacement_field<char,Handler&>(const Char *,const Char *,fmt::v9::detail::format_string_checker<Char,fmt::v9::detail::error_handler,unsigned __int64,std::basic_string<char,std::char_traits<char>,std::allocator<char>>>&)'
            with
            [
                Char=char,
                Handler=checker
            ]
    ...\fmt\core.h(2662): note: while evaluating function 'int fmt::v9::detail::format_string_checker<Char,fmt::v9::detail::error_handler,unsigned __int64,std::basic_string<char,std::char_traits<char>,std::allocator<char>>>::on_arg_id(void)'
            with
            [
                Char=char
            ]
    ...\fmt\core.h(2971): note: while evaluating function 'int fmt::v9::detail::compile_parse_context<Char,ErrorHandler>::next_arg_id(void)'
            with
            [
                Char=char,
                ErrorHandler=fmt::v9::detail::error_handler
            ]
    ...\fmt\core.h(747): note: while evaluating function 'void fmt::v9::basic_format_parse_context<char,fmt::v9::detail::error_handler>::on_error(const char *)'
```

**Comment by cyyever** [comment]
@clee2000 I restored the old behavior under FBCODE macro.

**Comment by cyyever** [comment]
@clee2000 The failures seem unrelated. If META builds work, then it's possible to merge.

**Comment by clee2000** [comment]
Yeah I'm still seeing stuff like 
```
    ...\Caffe2Common.cpp(42): error C2440: '<function-style-cast>': cannot convert from 'initializer list' to 'fmt::v9::string_view'
    ...\Caffe2Common.cpp(44): note: No constructor could take the source type, or constructor overload resolution was ambiguous

```

```
    ...\NNTorchScriptBackend.cpp(647): error C2440: '<function-style-cast>': cannot convert from 'initializer list' to 'fmt::v9::string_view'
    ...\NNTorchScriptBackend.cpp(647): note: 'fmt::v9::basic_string_view<char>::basic_string_view': no overloaded function could convert all the argument types
    ...\core.h(442): note: could be 'fmt::v9::basic_string_view<char>::basic_string_view(const Char *,size_t) noexcept'
            with
            [
                Char=char
            ]
    ...\NNTorchScriptBackend.cpp(647): note: 'fmt::v9::basic_string_view<char>::basic_string_view(const Char *,size_t) noexcept': cannot convert argument 1 from 'std::_String_view_iterator<_Traits>' to 'const Char *'
            with
            [
                Char=char
            ]
            and
            [
                _Traits=std::char_traits<char>
            ]
            and
            [
                Char=char
            ]
    ...\NNTorchScriptBackend.cpp(647): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
    ...\core.h(468): note: or       'fmt::v9::basic_string_view<char>::basic_string_view(S) noexcept'
    ...\core.h(462): note: or       'fmt::v9::basic_string_view<char>::basic_string_view(const std::basic_string<Char,Traits,Alloc> &) noexcept'
            with
            [
                Char=char
            ]
    ...\NNTorchScriptBackend.cpp(647): note: while trying to match the argument list '(std::_String_view_iterator<_Traits>, std::basic_string_view<char,std::char_traits<char>>::size_type)'
            with
            [
                _Traits=std::char_traits<char>
            ]
    ...\NNTorchScriptBackend.cpp(647): error C3615: constexpr function 'fmt::v9::basic_format_parse_context<char,fmt::v9::detail::error_handler>::on_error' cannot result in a constant expression
    ...\core.h(718): note: failure was caused by call of undefined function or one not declared 'constexpr'
    ...\core.h(718): note: see usage of 'fmt::v9::detail::error_handler::on_error'
    ...\NNTorchScriptBackend.cpp(647): note: the call stack of the evaluation (the oldest call first) is
    ...\NNTorchScriptBackend.cpp(647): note: while evaluating function 'fmt::v9::basic_format_string<char,const size_t &,std::basic_string<char,std::char_traits<char>,std::allocator<char>>>::basic_format_string<char[48],0>(const S (&))'
            with
            [
                S=char [48]
            ]
    ...\core.h(3169): note: while evaluating function 'void fmt::v9::detail::parse_format_string<true,char,checker>(fmt::v9::basic_string_view<char>,Handler &&)'
            with
            [
                Handler=checker
            ]
    ...\core.h(2731): note: while evaluating function 'const Char *fmt::v9::detail::parse_replacement_field<char,Handler&>(const Char *,const Char *,fmt::v9::detail::format_string_checker<Char,fmt::v9::detail::error_handler,unsigned __int64,std::basic_string<char,std::char_traits<char>,std::allocator<char>>>&)'
            with
            [
                Char=char,
                Handler=checker
            ]
    ...\core.h(2662): note: while evaluating function 'int fmt::v9::detail::format_string_checker<Char,fmt::v9::detail::error_handler,unsigned __int64,std::basic_string<char,std::char_traits<char>,std::allocator<char>>>::on_arg_id(void)'
            with
            [
                Char=char
            ]
    ...\core.h(2971): note: while evaluating function 'int fmt::v9::detail::compile_parse_context<Char,ErrorHandler>::next_arg_id(void)'
            with
            [
                Char=char,
                ErrorHandler=fmt::v9::detail::error_handler
            ]
    ...\core.h(747): note: while evaluating function 'void fmt::v9::basic_format_parse_context<char,fmt::v9::detail::error_handler>::on_error(const char *)'
```

This seems to be a windows only thing, is there a chance I need to change something internally as well to enable this?

Also seeing
```
*** Assertion failure in void FBRegisterAllGoogleTestCases()(), FBGoogleXCTestUtils.mm:192
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Failed to add Google Test method \"testConversionToString\", this method may already exist in the class.'
```
on mac/ios tests

File: c10/util/string_view.h [hunk 1]
```diff
@@ -588,8 +595,12 @@ constexpr inline void swap(
     basic_string_view<CharT>& rhs) noexcept {
   lhs.swap(rhs);
 }
-
+using string_view_ext = basic_string_view<char>;
+#ifdef fbcode_caffe2
```
Comment (malfet) [inline c10/util/string_view.h:599]: This should have been uppercase :)
```suggestion
#ifdef FBCODE_CAFFE2
```

File: c10/test/util/string_view_test.cpp [hunk 2]
```diff
@@ -66,6 +66,7 @@ static_assert(4 == hell.size(), "");
 static_assert(string_equal("hell", hell.data(), hell.size()), "");
 } // namespace test_sized_constructor
 
+#ifndef fbcode_caffe2
```
Comment (malfet) [inline c10/test/util/string_view_test.cpp:69]:
```suggestion
#ifndef FBCODE_CAFFE2
```

**Review by r-barnes — CHANGES_REQUESTED** [review]

**Review by r-barnes — APPROVED** [review]

File: c10/util/TypeIndex.h [hunk 3]
```diff
@@ -32,21 +32,21 @@ struct type_index final : IdWrapper<type_index, uint64_t> {
...
+      ? (throw std::logic_error("Invalid pattern"), string_view_ext())
       : str.substr(prefix.size(), str.size() - prefix.size() - suffix.size());
 #else
   return str.substr(prefix.size(), str.size() - prefix.size() - suffix.size());
 #endif
 }
 
 template <typename T>
-inline constexpr c10::string_view fully_qualified_type_name_impl() {
+inline constexpr c10::string_view_ext fully_qualified_type_name_impl() {
```
[the remark this answers is not in the document]
Comment (cyyever) [reply c10/util/TypeIndex.h:49]: Different c++ libraries will have different demangled name (libc++,libstdc++, MSVC STL).

File: caffe2/serialize/inline_container.cc [hunk 4]
```diff
@@ -283,7 +283,7 @@ size_t getPadding(
 bool PyTorchStreamReader::hasRecord(const std::string& name) {
   std::lock_guard<std::mutex> guard(reader_lock_);
 
-  if ((!load_debug_symbol_) && c10::string_view(name).ends_with(kDebugPklSuffix)) {
+  if ((!load_debug_symbol_) && c10::string_view_ext(name).ends_with(kDebugPklSuffix)) {
```
[the remark this answers is not in the document]
Comment (cyyever) [reply caffe2/serialize/inline_container.cc:286]: In a latter PR.

**Comment by malfet** [comment]
Just FYI it still fails internally, I guess parts of the code need to be guarded a bit further...

**Comment by cyyever** [comment]
> Just FYI it still fails internally, I guess parts of the code need to be guarded a bit further...

Should be the problem of old libfmt from the posted info.

**Pull request closed without merging**